postgres.go

```diff
@@ -0,0 +1,374 @@
+	ServiceName string `json:"service_name"`
+	Username    string `json:"username"`
+	Password    string `json:"password"`
+}
+
+func NewPostgresStore(c PostgresConfig) (*PostgresStore, error) {
+	var host string
+	if c.ServiceName != "" {
+		host = c.ServiceName + ".service.consul"
+	} else {
+		host = c.Addr
+	}
+	log.Infof("Connecting to postgres: %v", host)
+	psqlInfo := fmt.Sprintf("host=%s port=%d user=%s "+
+		"password=%s dbname=%s sslmode=disable",
+		host, c.Port, c.Username, c.Password, c.Database)
+	db, err := sql.Open("postgres", psqlInfo)
+	if err != nil {
+		return nil, errors.Wrap(err, "Error creating postgres session")
+	}
+	err = db.Ping()
+	if err != nil {
+		return nil, errors.Wrap(err, "Error creating postgres session")
+	}
+	log.Infof("Successfully connected to postgres %s", host)
+
+	return &PostgresStore{
+		db: db,
+	}, nil
+}
+
+func (p *PostgresStore) AddEvent(e *Event) error {
+	data := "{}"
+	if e.Data != nil {
+		dataBytes, err := json.Marshal(e.Data)
+		if err != nil {
+			return err
+		}
+		data = string(dataBytes)
+	}
+
+	tx, err := p.db.Begin()
+	if err != nil {
+		return err
+	}
+
+	_, err = tx.Exec("INSERT INTO event "+
+		"(event_id, parent_event_id, dc_id, topic_id, host, target_host_set, \"user\", event_time, tag_set, received_time)"+
+		" VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10)",
+		e.EventID, e.ParentEventID, e.DCID, e.TopicID, e.Host, pq.Array(e.TargetHosts), e.User, e.EventTime, pq.Array(e.Tags), e.ReceivedTime)
+
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+
+	_, err = tx.Exec("INSERT INTO event_metadata (event_id, data_json) VALUES ($1, $2)", e.EventID, data)
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+
+	return tx.Commit()
+}
+
+func getPlaceHolderArray(slice []string) string {
+	expr := "ARRAY["
+	first := true
+	for i := 0; i < len(slice); i++ {
+		if !first {
+			expr += ", "
+		}
+		expr += "?"
+		first = false
+	}
+	expr += "]"
+	return expr
+}
+
+func stringsToIfaces(in []string) []interface{} {
+	new := make([]interface{}, len(in))
+	for i, v := range in {
+		new[i] = v
+	}
+	return new
+}
+
+func (p *PostgresStore) Find(q *eventmaster.Query, topicIDs []string, dcIDs []string) (Events, error) {
+	// use sql builder for better readability
+	psql := sq.StatementBuilder.PlaceholderFormat(sq.Dollar)
+	columns := []string{"event_id", "parent_event_id", "dc_id", "topic_id", "host", "target_host_set", "\"user\"", "event_time", "tag_set", "received_time"}
+	table := "event"
+	base_query := psql.Select(columns...).From(table)
+
+	wheres := sq.And{}
+
+	// add constarints, *1000 due to different time precision (sec to ms)
+	wheres = append(wheres, sq.GtOrEq{"event_time": q.StartEventTime * 1000})
+	wheres = append(wheres, sq.LtOrEq{"event_time": q.EndEventTime * 1000})
+
+	if len(q.User) > 0 {
+		wheres = append(wheres, sq.Eq{"user": q.User})
+	}
+
+	if len(q.ParentEventID) > 0 {
+		wheres = append(wheres, sq.Eq{"parent_event_id": q.ParentEventID})
+	}
+
+	if len(q.Host) > 0 {
```

Comment:
@taoruicui For empty strings, whats the best way to check this in Go?

Reply:
usually just do `q.Host != ""` 
but this works too